### ISSUE TITLE
ValueObjects/Embeddables changes aren't correctly detected after committing

### ISSUE THREAD

**Issue by Thessi** [javers/javers#1372]
When deleting an embedded object from an entity and committing that entity, Javers doesn't pick up on those changes when querying for that data. The deleted fields in the embeddable aren't recognized or reflected changes in any way. Other changes in the entity are picked up as `TerminalValueChange`.

A second case is deleting an entity containing an (initialized) embeddable. Again, all fields are emitted as a `TerminalValueChange`, but the fields inside the embeddable aren't mentioned in the changes.

If I instead diff the entities (`Javers#compare()`), I get all expected changes.

**Steps To Reproduce**
Runnable test case is here:

https://github.com/Thessi/javers/blob/master/javers-core/src/test/groovy/org/javers/core/cases/CaseRemoveValueObjectReference.groovy

I implemented tests for both cases, and showed that a diff of the objects correctly recognizes all changes, whereas going through the commit doesn't.

**Javers' Version**
7.4.2

**Additional context**
I found some workarounds for both cases (in case someone urgently needs this functionality):

For the first case a workaround would be initializing all embeddables with empty objects, so Javers explicitly picks up on every null field. 

For the second case, the workaround is more difficult and depends on the typeof deletion. If we use `Javers#commitShallowDelete()`, we can first set all embeddable fields in the entity to an empty object, and persist it using `Javers#commit()`. Then, we commit the delete. Note: This will split the deletion in two separate commits. Also this only works if the deleted entity is merged and up to date with the entityManager. Otherwise this might also trigger some other changes.

If instead `javers#commitShallowDeleteById()` is used, and currently don't see a consistent workaround.

**Comment by bartoszwalacik** [comment]
Hi @Thessi , thanks for a high quality report, I will take a look on this

**Comment by bartoszwalacik** [comment]
Hi @Thessi , you correctly identified the issues.

The first case is a bug. A Value Object is a component of an owning Entity, so if a VO is nulled, it should be marked as deleted in Javers repo. Which means Javers should commit an explicit Terminal Snapshot for a nulled VO. Then, changes should be calculated correctly. Unfortunately, if we fix it this change won't be backward compatible, it will correct only new commits. Still, I believe it's important to fix it.

The second case is more like a missing feature than a bug, as the docs says:

```
    /**
     * Marks given object as deleted.
     * <br/><br/>
     *
     * Unlike {@link Javers#commit(String, Object)}, this method is shallow
     * and affects only given object.
     * <br/><br/>
     *
     * This method doesn't delete anything from JaVers repository.
     * It just persists 'terminal snapshot' of a given object.
     *
     * @param deleted object to be marked as deleted (Entity or Value Object)
     */
    Commit commitShallowDelete(String author, Object deleted);
```
Please note that VOs are committed in separate snapshots.
But since there is no good workaround to perform a cascade delete, I agree that a new method would come in handy.
We can name it `commitCascadeDelete()` -- 

```
    /**
     * Marks given and all its child Value Objects object as deleted.
     * <br/><br/>
     *  
     * This method doesn't delete anything from JaVers repository.
     * It just persists 'terminal snapshot' of a given object and its children.
     *
     * @param deleted object to be marked as deleted (Entity or Value Object)
     */
    Commit commitCascadeDelete(String author, Object deleted);
```

Please let me know if you are eager to contribute a PR

**Comment by Thessi** [comment]
I'll try to work on a PR in the next few days. I'll start with the first case for now, I'll let you know once I got something working.

**Comment by bartoszwalacik** [comment]
good luck @Thessi 👍

**Comment by olsavmic** [comment]
Hi @bartoszwalacik,
we have arrived to the same bug mentioned in this issue (ValueObject removal is not visible in the generated diff), yet when I looked into the code, it seems like this was partly by design:

`org.javers.core.JaversCommitE2ETest#should not record ObjectRemoved for removed ValueObject`

Do you remember why? I'd like to work on the fix but this seems intentional. Thank you!

**Comment by bartoszwalacik** [comment]
Good catch @olsavmic , indeed that test case stays in opposition. Still, seems that we all agree that this logic should be changed, so also this test should be aligned with the new logic.

**Comment by olsavmic** [comment]
I've created a draft PR with a fix for the TerminalValueChange on ValueObjects (https://github.com/javers/javers/pull/1430).

I'd appreciate some guidance (@bartoszwalacik), as I slightly doubt that adding the TerminalValueChange for VO conceptually fits the approach towards VOs in Javers 🤔

**Comment by bartoszwalacik** [comment]
Hi @olsavmic , I did some digging. You are right , that adding the TerminalValueChange for VO reference when this reference is unset is not the right direction. What we need is the set of TerminalValueChanges for all props of the unset VO.

It will require a bit more coding. The problem lies SnapshotDiffer.calculateDiffs(). Two facts are important:

- Javers doesn't generate TerminalSnapshots fo VO
- SnapshotDiffer composes the big diff from a list of atomic diffs when only two snapshots are compared, the new version with the old version.
- since there is no TerminalSnapshots fo VO, we miss one atomic diff -- TerminalSnapshots VO vs Previous Snapshots VO

So diff in findChanges() (which is based on SnapshotDiffer) works differently than diff in compare(), as in compare() only one big diff is calculated on the object graph level is calculated (no atomic one-to-one diffs!). And in this case, Javers can figure out which objects are removed from the new graph, and for them, this (virtual) Terminal vs Previous diff is calculated

**Comment by bartoszwalacik** [comment]
To solve the problem, I decided to introduce terminal snapshots for removed Value Objects. This change is not backward compatible

**Comment by bartoszwalacik** [comment]
the fix released in 7.9.0, see https://javers.org/release-notes

**Comment by alien11689** [comment]
Hi, I encountered this issue today and I am really happy that it was partially solved via the terminal snapshot. Thanks to it I was able to add the workaround to generate final diff not only based on the commit diff but also parsing terminal snapshots. It's definitely not perfect, but it partially solves the problem.

I do not know exactly the flow look like but I would suggest not setting `fixed & released` label on the partially solved issues since it is missleading.
